[hal,wpilibc,wpilibj,cmd] Update POV to use enums

KangarooKoala opened the pull request:
#7977 updated the meaning of the POV values, but did not update all of the usages. This PR switches everything to use the new HAL_JoystickPOV and DriverStation.POVDirection enums. In addition, it changes DriverStationJNI.setJoystickPOV to use a byte for the value parameter instead of an int to match the other POV JNI functions and adds a `HAL_ENUM_WITH_UNDERLYING_TYPE` macro.

ThadHouse commented:
Yeah we'll need to update generic hid too in other places. We should probably make an enum of some type.

KangarooKoala commented:
> Yeah we'll need to update generic hid too in other places.

Yep, based on grepping for `135` (and ignoring the many, many unrelated results), the other spots to update are the `GenericHID` method doc comments, the `CommandGenericHID` implementations, and variable names in `DriverStationGUI.cpp`.

> We should probably make an enum of some type.

Makes sense! Would we define the enum in [`hal/DriverStation.h`](https://github.com/wpilibsuite/allwpilib/blob/2027/hal/src/main/native/include/hal/DriverStation.h), [`frc/DriverStation.h`](https://github.com/wpilibsuite/allwpilib/blob/2027/wpilibc/src/main/native/include/frc/DriverStation.h), GenericHID, or elsewhere? Also, for C++ would we want an enum or an enum class?

KangarooKoala commented:
I decided to define the `HAL_JoystickPOV` enum in `hal/DriverStationTypes.h` so that the halsim_gui DriverStationGui.cpp implementation could reference it and define a `DriverStation.POVDirection` enum in `frc/DriverStation.h` for everything else to use so they don't need to interact with the HAL enum.

KangarooKoala commented:
The GenericHID and CommandGenericHID POV methods work in sim, based on using these Robot constructors in the developerRobot subproject:
```java
  public Robot() {
    var controller = new CommandGenericHID(0);
    var eventLoop = CommandScheduler.getInstance().getDefaultButtonLoop();

    // GenericHID
    controller.getHID().povUp(eventLoop).rising().ifHigh(() -> System.out.println("Up"));
    controller.getHID().povUpRight(eventLoop).rising().ifHigh(() -> System.out.println("UpRight"));
    controller.getHID().povRight(eventLoop).rising().ifHigh(() -> System.out.println("Right"));
    controller.getHID().povDownRight(eventLoop).rising().ifHigh(() -> System.out.println("DownRight"));
    controller.getHID().povDown(eventLoop).rising().ifHigh(() -> System.out.println("Down"));
    controller.getHID().povDownLeft(eventLoop).rising().ifHigh(() -> System.out.println("DownLeft"));
    controller.getHID().povLeft(eventLoop).rising().ifHigh(() -> System.out.println("Left"));
    controller.getHID().povUpLeft(eventLoop).rising().ifHigh(() -> System.out.println("UpLeft"));
    controller.getHID().povCenter(eventLoop).rising().ifHigh(() -> System.out.println("Center"));

    // CommandGenericHID
    controller.povUp().onTrue(Commands.print("cmd Up"));
    controller.povUpRight().onTrue(Commands.print("cmd UpRight"));
    controller.povRight().onTrue(Commands.print("cmd Right"));
    controller.povDownRight().onTrue(Commands.print("cmd DownRight"));
    controller.povDown().onTrue(Commands.print("cmd Down"));
    controller.povDownLeft().onTrue(Commands.print("cmd DownLeft"));
    controller.povLeft().onTrue(Commands.print("cmd Left"));
    controller.povUpLeft().onTrue(Commands.print("cmd UpLeft"));
    controller.povCenter().onTrue(Commands.print("cmd Center"));

    addPeriodic(() -> CommandScheduler.getInstance().run(), kDefaultPeriod);
  }
```

```cpp
  frc2::CommandGenericHID controller{0};

 public:
  Robot() {
    frc::EventLoop* eventLoop =
        frc2::CommandScheduler::GetInstance().GetDefaultButtonLoop();

    // GenericHID
    controller.GetHID().POVUp(eventLoop).Rising().IfHigh(
        [] { wpi::print("Up\n"); });
    controller.GetHID().POVUpRight(eventLoop).Rising().IfHigh(
        [] { wpi::print("UpRight\n"); });
    controller.GetHID().POVRight(eventLoop).Rising().IfHigh(
        [] { wpi::print("Right\n"); });
    controller.GetHID().POVDownRight(eventLoop).Rising().IfHigh(
        [] { wpi::print("DownRight\n"); });
    controller.GetHID().POVDown(eventLoop).Rising().IfHigh(
        [] { wpi::print("Down\n"); });
    controller.GetHID().POVDownLeft(eventLoop).Rising().IfHigh(
        [] { wpi::print("DownLeft\n"); });
    controller.GetHID().POVLeft(eventLoop).Rising().IfHigh(
        [] { wpi::print("Left\n"); });
    controller.GetHID().POVUpLeft(eventLoop).Rising().IfHigh(
        [] { wpi::print("UpLeft\n"); });
    controller.GetHID().POVCenter(eventLoop).Rising().IfHigh(
        [] { wpi::print("Center\n"); });

    // CommandGenericHID
    controller.POVUp().OnTrue(frc2::cmd::Print("cmd Up"));
    controller.POVUpRight().OnTrue(frc2::cmd::Print("cmd UpRight"));
    controller.POVRight().OnTrue(frc2::cmd::Print("cmd Right"));
    controller.POVDownRight().OnTrue(frc2::cmd::Print("cmd DownRight"));
    controller.POVDown().OnTrue(frc2::cmd::Print("cmd Down"));
    controller.POVDownLeft().OnTrue(frc2::cmd::Print("cmd DownLeft"));
    controller.POVLeft().OnTrue(frc2::cmd::Print("cmd Left"));
    controller.POVUpLeft().OnTrue(frc2::cmd::Print("cmd UpLeft"));
    controller.POVCenter().OnTrue(frc2::cmd::Print("cmd Center"));

    AddPeriodic([] { frc2::CommandScheduler::GetInstance().Run(); },
                kDefaultPeriod);
  }
```

KangarooKoala commented:
@ThadHouse Could you review please?

ThadHouse commented on wpilibj/src/main/java/edu/wpi/first/wpilibj/DriverStation.java:
```diff
@@ -89,6 +89,51 @@ public enum MatchType {
   }
 
+  /** A controller POV direction. */
+  public enum POVDirection {
```
So I'm ok adding it, so its good for now, however in general its going to be useless. Because of the requirement for a HW DS, joysticks will likely be switching to SDL. SDL doesn't actually use POVs for gamepads, and instead maps the dpad on gamepads to buttons directly. So users would have to manually construct a POVDirection.

KangarooKoala replied:
Ah, good to know about the likely switch to SDL! I'm logging off soon, but I'll make the suggested changes tomorrow. (Move `getAngle()` to a `POVDirection` non-static method, switch to `default` for the C++ impl, and avoid throwing in `POVDirection.of()`)

ThadHouse commented on wpilibc/src/main/native/include/frc/DriverStation.h:
```diff
@@ -46,6 +49,60 @@ class DriverStation final {
+   */
+  static constexpr std::optional<Rotation2d> GetAngle(POVDirection angle) {
+    switch (angle) {
+      case kCenter:
```
`default` instead of `case kCenter`. In C++, this actually won't compile as not all cases return a value.

ThadHouse commented on wpilibj/src/main/java/edu/wpi/first/wpilibj/DriverStation.java:
```diff
@@ -89,6 +90,72 @@ public enum MatchType {
+          return direction;
+        }
+      }
+      throw new IllegalArgumentException("Invalid value " + value + "!");
```
We should make invalid values be center and print an error, rather than throwing. There have been controllers in the past that have allowed 3 buttons, and they could possibly crash. I'm not sure I know what SDL does and if I should trust it.

KangarooKoala commented:
Could people from wpilibsuite/commandbased and/or wpilibsuite/simulation review this please?

ThadHouse commented:
We're waiting until after the initial alpha release to merge this. Want to delay the break for just a bit.

KangarooKoala commented:
Rebased and squashed onto 2027. Diff should be the same as it was previously- Prior to pushing, `git diff origin/generic-hid-update > a.patch; git diff $(git merge-base 2027 origin/generic-hid-update) 2027 > b.patch; git diff --no-index a.patch b.-patch` had no output.
